### ISSUE TITLE
Comparison Operator: Values of types int and float with same numerical value are equal

### ISSUE THREAD

**Issue by nanoquack** [godotengine/godot#80162]
### Godot version

v4.1.stable.official [970459615]

### System information

Windows 11 Pro, Version 22H2, Build 22621.1778 - v4.1.stable.official [9704596] - Vulkan (Forward +)

### Issue description

It's not really a bug, but it is somehow counterintuitive: While two variables can have different types int and float, they are treated as equal by the comparison operator:
```
var val1 = 42
var val2 = 42.0
print(val1 == val2) #true
print(typeof(val1) == typeof(val2)) #false
```

This gets especially interesting, if you add an int and a float to an array or a dictionary and compare them:
```
var val1 = [42]
var val2 = [42.0]
print(val1 == val2) #false
print(val1[0] == val2[0]) #true
```
As you can see, the two arrays are not considered equal, and rightly so, but the two values are considered equal. Wouldn't it be much more logical if comparison of an int and a float always returns false?

### Steps to reproduce

See above

### Minimal reproduction project

N/A

**Comment by AThousandShips** [comment]
The first behavior is not going to change, pretty certainly, it's pretty universal across different languages to allow this, as it is a major source of confusion and errors

The second part is because when checking equality of `Array` it checks the type as well, this should be documented because it can be confusing

**Comment by dalexeev** [comment]
See:

* #26375
* #31331
* godotengine/godot-proposals#4571
* godotengine/godot-proposals#5101
* godotengine/godot-proposals#6092
* #71758

[GDScript reference - Operators](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_basics.html#operators):

![](https://github.com/godotengine/godot/assets/47700418/43133042-733e-48d6-8502-a94951d83f2d)

**Comment by nanoquack** [comment]
I agree that many languages do it this way (especially more dynamically typed ones), and I didn't mean to imply the behaviour has to be changed, but I think as @AThousandShips proposed the behaviour for Array and Dictionary should be documented
